common/functions/src/scalars/maths/log.rs

```diff
@@ -14,153 +14,134 @@
+    fn base() -> f64 {
+        10f64
     }
 }
 
-impl fmt::Display for GenericLogFunction {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}", self.display_name.to_uppercase())
+impl Base for TwoBase {
+    fn base() -> f64 {
+        2f64
     }
 }
+#[derive(Clone)]
+pub struct GenericLogFunction<T> {
+    display_name: String,
+    t: PhantomData<T>,
+}
 
-pub struct LogFunction {}
-
-impl LogFunction {
-    pub fn try_create(display_name: &str) -> Result<Box<dyn Function>> {
-        GenericLogFunction::try_create(display_name, E)
+impl<T: Base> GenericLogFunction<T> {
+    pub fn try_create(display_name: &str) -> Result<Box<dyn Function2>> {
+        Ok(Box::new(Self {
+            display_name: display_name.to_string(),
+            t: PhantomData,
+        }))
     }
 
-    pub fn desc() -> FunctionDescription {
-        FunctionDescription::creator(Box::new(Self::try_create)).features(
+    pub fn desc() -> Function2Description {
+        Function2Description::creator(Box::new(Self::try_create)).features(
             FunctionFeatures::default()
                 .deterministic()
                 .variadic_arguments(1, 2),
         )
     }
-}
 
-pub struct LnFunction {}
-
-impl LnFunction {
-    pub fn try_create(display_name: &str) -> Result<Box<dyn Function>> {
-        GenericLogFunction::try_create(display_name, E)
+    fn log<S>(value: S) -> f64
+    where S: AsPrimitive<f64> {
+        value.as_().log(T::base())
     }
 
-    pub fn desc() -> FunctionDescription {
-        FunctionDescription::creator(Box::new(Self::try_create))
-            .features(FunctionFeatures::default().deterministic().num_arguments(1))
+    fn log_with_base<S, B>(base: S, value: B, _ctx: &mut EvalContext) -> f64
+    where
+        S: AsPrimitive<f64>,
+        B: AsPrimitive<f64>,
+    {
+        value.as_().log(base.as_())
     }
 }
 
-pub struct Log10Function {}
-
-impl Log10Function {
-    pub fn try_create(display_name: &str) -> Result<Box<dyn Function>> {
-        GenericLogFunction::try_create(display_name, 10_f64)
+impl<T: Base> Function2 for GenericLogFunction<T> {
+    fn name(&self) -> &str {
+        &*self.display_name
     }
 
-    pub fn desc() -> FunctionDescription {
-        FunctionDescription::creator(Box::new(Self::try_create))
-            .features(FunctionFeatures::default().deterministic().num_arguments(1))
+    fn return_type(&self, args: &[&DataTypePtr]) -> Result<DataTypePtr> {
+        for arg in args {
+            assert_numeric(*arg)?;
+        }
+        Ok(f64::to_data_type())
     }
-}
 
-pub struct Log2Function {}
-
-impl Log2Function {
-    pub fn try_create(display_name: &str) -> Result<Box<dyn Function>> {
-        GenericLogFunction::try_create(display_name, 2_f64)
+    fn eval(&self, columns: &ColumnsWithField, _input_rows: usize) -> Result<ColumnRef> {
+        if columns.len() == 1 {
+            with_match_primitive_type_id!(columns[0].data_type().data_type_id(), |$S| {
```

Comment:
Maybe use with_match_date_type_error is better

Reply:
It does not matter, we already checked types must be numeric in `return_type`.